SteamKit2/SteamKit2/Util/Win32Helpers.cs

```diff
@@ -1,258 +1,168 @@
-				public uint VendorIdOffset;
-				public uint ProductIdOffset;
-				public uint ProductRevisionOffset;
-				public uint SerialNumberOffset;
-				public int StorageBusType;
-				public uint RawPropertiesLength;
-				[MarshalAs( UnmanagedType.ByValArray, SizeConst = 1 )]
-				public byte[] RawDeviceProperties;
-			}
-
-#pragma warning restore 0649
-
-			public static uint IOCTL_VOLUME_GET_VOLUME_DISK_EXTENTS = 0x00560000;
-			public static uint IOCTL_STORAGE_QUERY_PROPERTY = 0x002D1400;
-			public static int StorageDeviceProperty = 0;
-			public static int PropertyStandardQuery = 0;
-
-			[LibraryImport( "kernel32.dll", SetLastError = true )]
-			[return: MarshalAs( UnmanagedType.Bool )]
-			public static partial bool DeviceIoControl( SafeHandle hDevice, uint dwIoControlCode, IntPtr lpInBuffer, uint nInBufferSize, IntPtr lpOutBuffer, uint nOutBufferSize, out uint lpBytesReturned, IntPtr lpOverlapped );
-
-			#endregion
-		}
-	}
+    [SupportedOSPlatform( "windows5.1.2600" )]
+    static partial class Win32Helpers
+    {
+        public static string? GetBootDiskSerialNumber()
+        {
+            try
+            {
+                var bootDiskNumber = GetBootDiskNumber();
+                var serialNumber = GetPhysicalDriveSerialNumber( bootDiskNumber );
+                return serialNumber;
+            }
+            catch
+            {
+                return null;
+            }
+        }
+
+        static string GetBootDiskLogicalVolume()
+        {
+            var volume = Path.GetPathRoot( Environment.SystemDirectory.AsSpan() ).TrimEnd( Path.DirectorySeparatorChar );
+
+            if ( volume.Length == 0 )
+            {
+                throw new InvalidOperationException( "Could not determine system drive letter." );
+            }
+
+            return volume.ToString();
+        }
+
+        static unsafe uint GetBootDiskNumber()
+        {
+            using var handle = PInvoke.CreateFile(
+                $@"\\.\{GetBootDiskLogicalVolume()}",
+                0,
+                FILE_SHARE_MODE.FILE_SHARE_READ | FILE_SHARE_MODE.FILE_SHARE_WRITE,
+                null,
+                FILE_CREATION_DISPOSITION.OPEN_EXISTING,
+                FILE_FLAGS_AND_ATTRIBUTES.FILE_ATTRIBUTE_NORMAL,
+                null
+            );
+
+            if ( handle == null || handle.IsInvalid )
+            {
+                throw new FileNotFoundException( "Unable to open boot volume." );
+            }
+
+            var extents = new VOLUME_DISK_EXTENTS();
+
+            if ( !PInvoke.DeviceIoControl(
+                handle,
+                PInvoke.IOCTL_VOLUME_GET_VOLUME_DISK_EXTENTS,
+                null,
+                0,
+                &extents,
+                ( uint )VOLUME_DISK_EXTENTS.SizeOf( 1 ),
+                null,
+                null
+            ) )
+            {
+                throw new InvalidOperationException( $"Failed to get volume disk extents: {Marshal.GetLastWin32Error()}" );
```

Comment:
Why not just Win32Exception? It already includes the last Win32 error and the stack trace should tell us where it went wrong.

Reply:
It does? I saw dotnet code has an internal code to throw win32 errors after invokes, but didn't check this exception.

Reply:
https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/ComponentModel/Win32Exception.cs,26